core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -0,0 +1,185 @@
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      _       <- buckets.update(idx, _.filterNot(_._1 == k))
+    } yield ()
+
+  /**
+   * Atomically folds using pure function.
+   */
+  final def fold[A](zero: A)(op: (A, (K, V)) => A): STM[Nothing, A] =
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.fold(zero)((acc, chain) => chain.foldLeft(acc)(op))
+    } yield res
+
+  /**
+   * Atomically folds using effectful function.
+   */
+  final def foldM[A, E](zero: A)(op: (A, (K, V)) => STM[E, A]): STM[E, A] = {
+    def loopM(acc: STM[E, A], remaining: List[(K, V)]): STM[E, A] =
+      remaining match {
+        case Nil          => acc
+        case head :: tail => loopM(acc.flatMap(op(_, head)), tail)
+      }
+
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.foldM(zero)((acc, chain) => loopM(STM.succeed(acc), chain))
+    } yield res
+  }
+
+  /**
+   * Atomically performs side-effect for each binding present in map.
+   */
+  final def foreach[E](f: ((K, V)) => STM[E, Unit]): STM[E, Unit] =
+    self.foldM(())((_, kv) => f(kv))
+
+  /**
+   * Retrieves value associated with given key.
+   */
+  final def get[E](k: K): STM[E, Option[V]] =
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      bucket  <- buckets(idx)
+    } yield bucket.find(_._1 == k).map(_._2)
+
+  /**
+   * Retrieves value associated with given key or default value, in case the
+   * key isn't present.
+   */
+  final def getOrElse[E](k: K, default: => V): STM[E, V] =
+    get(k).map(_.getOrElse(default))
+
+  /**
+   * Creates new [[TMap]] by mapping all bindings using pure function.
+   */
+  final def map[K2, V2](f: ((K, V)) => (K2, V2)): STM[Nothing, TMap[K2, V2]] =
+    self.fold(List.empty[(K2, V2)])((acc, kv) => f(kv) :: acc).flatMap(TMap(_))
+
+  /**
+   * Creates new [[TMap]] by mapping all bindings using effectful function.
+   */
+  final def mapM[E, K2, V2](f: ((K, V)) => STM[E, (K2, V2)]): STM[E, TMap[K2, V2]] =
+    self.foldM(List.empty[(K2, V2)])((acc, kv) => f(kv).map(_ :: acc)).flatMap(TMap(_))
+
+  /**
+   * Stores new binding into the map.
+   */
+  final def put[E](k: K, v: V): STM[E, Unit] = {
+    def upsert(bucket: List[(K, V)]): List[(K, V)] =
+      bucket match {
+        case Nil => List(k -> v)
+        case xs  => xs.map(kv => if (kv._1 == k) (k, v) else kv)
+      }
+
+    def resize(capacity: Int, data: List[(K, V)]): STM[Nothing, Unit] = {
+      val newCapacity = capacity * 2
+
+      for {
+        tmap       <- allocate(newCapacity, data)
+        newBuckets <- tmap.tBuckets.get
+        _          <- self.tBuckets.set(newBuckets)
+        _          <- self.tCapacity.set(newCapacity)
+      } yield ()
+    }
+
+    for {
+      buckets     <- tBuckets.get
+      idx         <- indexOf(k)
+      _           <- buckets.update(idx, upsert)
+      data        <- self.fold(List.empty[(K, V)])((acc, kv) => kv :: acc)
```

Comment:
Doesn't this grab the whole contents of the map, even if there is no need to resize? We should aim for maximum laziness to avoid doing unnecessary worm

Reply:
Good point, stored the size in `TRef`.